File: python/zimtohrli_test.py
# Copyright 2022 The Zimtohrli Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for google3.third_party.zimtohrli.python.zimtohrli."""

import unittest
import jax
import numpy as np
import audio_signal
import zimtohrli


class ZimtohrliTest(unittest.TestCase):

    def test_zimtohrli_spectrogram_and_distance(self):
        sample_rate = 48000.0
        signal_a = np.zeros((int(sample_rate) // 100,))
        signal_a[0] = 1.0
        signal_b = np.zeros((int(sample_rate) // 100,))
        signal_b[0:1] = 0.9
        sound_a = audio_signal.Signal(sample_rate=sample_rate, samples=signal_a)
        sound_b = audio_signal.Signal(sample_rate=sample_rate, samples=signal_b)

        def compute_distance(s_a, s_b):
            z = zimtohrli.Zimtohrli()
            spectrogram_a = z.spectrogram(s_a)
            spectrogram_b = z.spectrogram(s_b)
            distance = z.distance(spectrogram_a, spectrogram_b)
            return distance

        # Run it without jit
        distance = compute_distance(sound_a, sound_b)
        self.assertGreater(distance, 0)

        # Run it under jit
        jit_compute_distance = jax.jit(compute_distance)
        jit_distance = jit_compute_distance(sound_a, sound_b)
        self.assertGreater(jit_distance, 0)

        self.assertAlmostEqual(distance, jit_distance, delta=1e-5)


if __name__ == "__main__":
    unittest.main()

File: cpp/zimt/zimtohrli_test.cc
// Copyright 2024 The Zimtohrli Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "zimt/zimtohrli.h"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdlib>
#include <optional>
#include <utility>
#include <vector>

#include "benchmark/benchmark.h"
#include "gtest/gtest.h"

namespace zimtohrli {

namespace {

TEST(Zimtohrli, NormalizeAmplitudeTest) {
  std::vector<float> reference = {0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5};
  std::vector<float> signal = {0.25, 0.25, 0.25, 0.25, 0.25};
  const EnergyAndMaxAbsAmplitude reference_measurements = Measure(reference);
  EXPECT_NEAR(reference_measurements.energy_db_fs, 20 * std::log10(0.5 * 0.5),
              1e-4);
  EXPECT_EQ(reference_measurements.max_abs_amplitude, 0.5);
  const EnergyAndMaxAbsAmplitude signal_measurements =
      NormalizeAmplitude(reference_measurements.max_abs_amplitude, signal);
  for (size_t index = 0; index < signal.size(); ++index) {
    EXPECT_EQ(signal[index], 0.5);
  }
  EXPECT_NEAR(signal_measurements.energy_db_fs, 20 * std::log10(0.5 * 0.5),
              1e-4);
  EXPECT_EQ(signal_measurements.max_abs_amplitude, 0.5);
}

}  // namespace

}  // namespace zimtohrli
